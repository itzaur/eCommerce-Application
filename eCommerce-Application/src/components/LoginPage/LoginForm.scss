@font-face {
    font-family: 'Neuropol';
    src: url(../../assets/fonts/neuropol/neuropol.otf);
}
@font-face {
    font-family: 'Bowler';
    src: url(../../assets/fonts/bowler/BOWLER.otf);
}
body,
section,
div,
h1,
h2,
div,
p,
button {
    margin: 0;
    padding: 0;
}
:root {
    --color-dark-l: #5e5971;
    --color-dark-m: #5e597180;
    --color-primary: #fff;
    --color-primary-opacity: #f5f5f5cc;
    --main-font: 'Neuropol';
    --second-font: 'Bowler';
    --color-button: #4de1e3;
}

html {
    font-size: 10px;
}
body {
    display: flex;
    align-items: center;

    background-image: url('../../assets/images/backround-login.png');
    background-size: cover;

    font-family: 'Neuropol', sans-serif;
}
#root {
    width: 100%;
}
.login {
    width: 47.4%;
    margin: 0 auto;

    font-family: 'Neuropol', sans-serif;
}
.logo-big {
    display: block;
    width: 26%;
    margin: 0 auto;
}
.form {
    border-radius: 2px;
}
.form__content {
    background-color: var(--color-primary-opacity);
}
.form__title {
    background-color: var(--color-dark-l);
    color: var(--color-primary);
}
.form__content {
    padding: 5.8% 9.3% 7.14% 10%;
}
.form__title {
    padding: 2.9% 0;

    font-size: clamp(1rem, 1.5vw, 2.8rem);
    text-align: center;
    line-height: 92%;
}
.form__question {
    margin-bottom: 7%;

    color: var(--color-dark-l);

    font-size: clamp(1rem, 1.04vw, 2rem);
    line-height: 92%;
    a {
        color: var(--color-dark-l);
        border-bottom: dotted #5e5971;
    }
}

.form__inputs {
    display: flex;
    flex-direction: column;
}

.placeinput input:invalid + .place_holder {
    display: inline;
}

.placeinput {
    position: relative;
    overflow: hidden;
    width: 100%;
    height: 5rem;
    input {
        position: absolute;
        z-index: 2;

        width: 97%;
        padding: 1rem 0 1rem 1rem;

        background-color: transparent;
        color: var(--color-dark-m);
        border: 1px solid var(--color-dark-l);
        outline: none;

        font-size: clamp(1rem, 1.04vw, 2rem);
        font-family: 'Bowler', sans-serif;
        &:focus {
            background-color: var(--color-primary);
            color: #000;
        }
    }
    .form__input_invalid {
        border: 1px solid rgb(255, 54, 54);
    }
}

.place_holder {
    display: none;

    position: absolute;
    top: 0;
    left: 0;
    z-index: 1;

    padding: 1rem;

    cursor: text;

    color: var(--color-dark-m);
    font-size: clamp(1rem, 1.04vw, 2rem);
    span {
        color: var(--color-button);
    }
}

.button-action {
    align-self: center;
    padding: 2rem 7.5rem;
    border-radius: 2px;

    background-color: var(--color-button);
    border: 2px solid var(--color-button);

    font-size: clamp(1rem, 1.04vw, 2rem);
    font-family: 'Bowler', sans-serif;
    &:focus {
        outline: none;
        border: 2px solid var(--color-button);
    }
    &:hover {
        border: none;
        border: 2px solid var(--color-button);
    }
}
.password-view,
.password-view:focus-visible,
.password-view:focus {
    position: absolute;
    right: 4%;
    height: inherit;
    z-index: 3;

    background-color: transparent;
    border: none;
    outline: none;
    background-size: contain;
    background-repeat: no-repeat;
}
.password-view_text {
    width: 4.1rem;
    top: 0.7rem;

    background-image: url('../../assets/images/eye-closed.png');
}
.password-view_password {
    width: 3.9rem;
    top: 1.02rem;

    background-image: url('../../assets/imgages/eyes-open.png');
}

.error-message {
    margin-bottom: 4.6rem;
    height: 1rem;

    color: rgb(255, 54, 54);
    font-size: 1.3rem;
}
