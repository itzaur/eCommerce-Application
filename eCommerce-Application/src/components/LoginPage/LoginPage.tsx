import { useState } from 'react';
import logo from '../../assets/images/logo.png';
import { checkIncorrectEmail } from '../../utils/validation/checkCorrectEmail';
import { checkIncorrectPassword } from '../../utils/validation/checkPassword';
import { loginCustomer } from '../../commercetools/loginCustomer';

// let user = null;

function LoginPage(): JSX.Element {
    const [errorEmail, setErrorEmail] = useState({ error: false, message: '' });
    const [errorPassword, setErrorPassword] = useState({
        error: false,
        message: '',
    });
    const [passwordView, setPasswordView] = useState('password');
    const [password, setPassword] = useState('');
    const [email, setEmail] = useState('');
    const [errorWithLogin, setErrorWithLogin] = useState(false);

    const changePasswordView = (): void => {
        setPasswordView(passwordView === 'password' ? 'text' : 'password');
    };

    const loginUser = (
        e: React.MouseEvent<HTMLButtonElement, MouseEvent>
    ): void => {
        e.preventDefault();
        if (!email) {
            setErrorEmail({ error: true, message: 'Это обязательное поле' });
        }
        if (!password) {
            setErrorPassword({
                error: true,
                message: 'Это обязательное поле',
            });
        }
        if (errorEmail.error || errorPassword.error || !email || !password)
            return;
        loginCustomer(email, password)
            .then(() => {
                setErrorWithLogin(false);
            })
            .catch((err) => {
                setErrorWithLogin(true);
                if (err.cause === 'passwordError') {
                    setErrorPassword({ error: true, message: err.message });
                } else if (err.cause === 'emailError') {
                    setErrorEmail({ error: true, message: err.message });
                }
            });
    };

    return (
        <>
            <img className="logo-big" src={logo} alt="logo" />
            <section className="form login">
                <h2 className="form__title">
                    Добро пожаловать <br /> На Борт Космической Одиссеи!
                </h2>
                <div className="form__content">
                    <div className="form__question">
                        Путешествуете с нами впервые?&nbsp;
                        <a href="##">Зарегистрируйтесь!</a>
                    </div>
                    <form className="form__inputs">
                        <div className="placeinput">
                            <input
                                className={
                                    errorEmail.error
                                        ? 'form__input form__input_invalid'
                                        : 'form__input'
                                }
                                type="text"
                                id="email"
                                required
                                onBlur={(e): void => {
                                    setErrorEmail({
                                        error: checkIncorrectEmail(e).incorrect,
                                        message: checkIncorrectEmail(e).message,
                                    });
                                }}
                                onChange={(e): void => {
                                    if (errorWithLogin) {
                                        setErrorPassword({
                                            error: false,
                                            message: '',
                                        });
                                    }
                                    setErrorEmail({
                                        error: checkIncorrectEmail(e, true)
                                            .incorrect,
                                        message: checkIncorrectEmail(e, true)
                                            .message,
                                    });
                                    setEmail(e.target.value);
                                }}
                            />
                            <div className="place_holder">
                                Email<span>*</span>
                            </div>
                        </div>
                        <p className="error-messages">
                            {errorEmail ? errorEmail.message : ''}
                        </p>
                        <div className="placeinput">
                            <input
                                className={
                                    errorPassword.error
                                        ? 'form__input form__input_invalid'
                                        : 'form__input'
                                }
                                type={passwordView}
                                id="password"
                                required
                                onBlur={(e): void => {
                                    setErrorPassword({
                                        error: checkIncorrectPassword(e)
                                            .incorrect,
                                        message:
                                            checkIncorrectPassword(e).message,
                                    });
                                }}
                                onChange={(e): void => {
                                    setErrorPassword({
                                        error: checkIncorrectPassword(e, true)
                                            .incorrect,
                                        message: checkIncorrectPassword(e, true)
                                            .message,
                                    });
                                    setPassword(e.target.value);
                                }}
                            />
                            <div className="place_holder">
                                Password<span>*</span>
                            </div>
                            <button
                                type="button"
                                className={`password-view password-view_${passwordView}`}
                                onClick={(): void => changePasswordView()}
                                onKeyDown={(): void => changePasswordView()}
                            >
                                &nbsp;
                            </button>
                        </div>
                        <p className="error-message">
                            {errorPassword ? errorPassword.message : ''}
                        </p>
                        <button
                            className="button-action"
                            type="submit"
                            onClick={(e): void => {
                                loginUser(e);
                            }}
                        >
                            Войти
                        </button>
                    </form>
                </div>
            </section>
        </>
    );
}

export default LoginPage;
