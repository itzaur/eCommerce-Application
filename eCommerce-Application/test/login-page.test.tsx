import {
    fireEvent,
    render,
    screen,
    RenderResult,
} from '@testing-library/react';
import { expect, describe, it } from 'vitest';
import React from 'react';
import LoginPage from '../src/components/LoginPage/LoginPage';
import { BrowserRouter } from 'react-router-dom';
import '@testing-library/jest-dom/extend-expect';

let page: null | RenderResult<
    typeof import('../node_modules/@testing-library/dom/types/queries'),
    HTMLElement,
    HTMLElement
> = null;

describe('Login tests', () => {
    beforeEach(() => {
        page = render(
            <BrowserRouter>
                <LoginPage />
            </BrowserRouter>
        );
    });
    it('Login page renders', () => {
        expect(screen.getByText(/Добро пожаловать/)).toHaveClass('form__title');
        expect(page).toMatchSnapshot();
    });
    it('Client validation of email works correctly', () => {
        expect(document.getElementById('email')).toBeInTheDocument();
        expect(
            screen.queryByText(/Введите e-mail в верном формате/)
        ).toBeNull();
        expect(screen.queryByDisplayValue(/cat/)).toBeNull();
        fireEvent.change(document.getElementById('email') as HTMLInputElement, {
            target: { value: 'cat' },
        });
        expect(screen.queryByDisplayValue(/cat/)).toBeInTheDocument();
        expect(
            screen.queryByText(/Введите e-mail в верном формате/)
        ).toBeInTheDocument();
    });
    it('Client validation of password works correctly', () => {
        expect(document.getElementById('password')).toBeInTheDocument();
        expect(
            screen.queryByText(/Пароль должен быть не менее 8 символов/)
        ).toBeNull();
        expect(screen.queryByDisplayValue(/pass/)).toBeNull();
        fireEvent.change(
            document.getElementById('password') as HTMLInputElement,
            {
                target: { value: 'pass' },
            }
        );
        expect(screen.queryByDisplayValue(/pass/)).toBeInTheDocument();
        expect(
            screen.queryByText(/Пароль должен быть не менее 8 символов/)
        ).toBeInTheDocument();
        fireEvent.change(
            document.getElementById('password') as HTMLInputElement,
            {
                target: { value: 'password' },
            }
        );
        expect(
            screen.queryByText(
                /Пароль должен содержать заглавную букву в диапазоне A - Z/
            )
        ).toBeInTheDocument();
        fireEvent.change(
            document.getElementById('password') as HTMLInputElement,
            {
                target: { value: 'passwordA1' },
            }
        );
        expect(
            screen.queryByText(
                'Пароль должен содержать специальный символ !@#$%^&*+'
            )
        ).toBeInTheDocument();
    });
    it('Correctly checks fields are empty', () => {
        expect(screen.queryByText(/Это обязательное поле/)).toBeNull();
        fireEvent.click(
            document.getElementById('button-login') as HTMLInputElement
        );
        expect(screen.queryAllByText(/Это обязательное поле/)).not.toBeNull();
    });
    it('Correctly checks correct password', () => {
        expect(screen.queryByText(/Вы ввели неверный пароль/)).toBeNull();
        fireEvent.change(document.getElementById('email') as HTMLInputElement, {
            target: { value: 'ksu1ven@mail.ru' },
        });
        fireEvent.change(
            document.getElementById('password') as HTMLInputElement,
            {
                target: { value: '9845587Ok+' },
            }
        );
        fireEvent.click(
            document.getElementById('button-login') as HTMLInputElement
        );

        expect(screen.queryByText(/Вы ввели неверный пароль/)).toBeNull();
    });
});
