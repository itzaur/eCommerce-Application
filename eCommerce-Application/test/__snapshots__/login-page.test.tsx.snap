// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Login tests > Login page renders 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <header
        class="header"
      >
        <a
          href="/"
        >
          <img
            alt="logo"
            class="logo_big"
            src="/src/assets/images/logo.png"
          />
        </a>
      </header>
      <section
        class="form login"
      >
        <h2
          class="form__title"
        >
          Добро пожаловать 
          <br />
           На Борт Космической Одиссеи!
        </h2>
        <div
          class="form__content"
        >
          <div
            class="form__question"
          >
            Путешествуете с нами впервые? 
            <a
              href="/registration"
            >
              Зарегистрируйтесь!
            </a>
          </div>
          <form
            class="form__inputs"
          >
            <div
              class="form__inputs-wrapper"
            >
              <label
                class="placeholder"
                for="name"
              >
                <input
                  class="form__input"
                  id="email"
                  required=""
                  type="text"
                />
                <div
                  class="placeholder__input form_big-first-letter"
                >
                  Email
                  <span>
                    *
                  </span>
                </div>
                <p
                  class="error-message "
                />
              </label>
            </div>
            <div
              class="form__inputs-wrapper"
            >
              <label
                class="placeholder"
                for="password"
              >
                <input
                  class="form__input"
                  id="password"
                  required=""
                  type="password"
                />
                <div
                  class="placeholder__input form_big-first-letter"
                >
                  Password
                  <span>
                    *
                  </span>
                </div>
                <button
                  class="password-view password-view_password"
                  type="button"
                >
                   
                </button>
                <p
                  class="error-message"
                />
              </label>
            </div>
            <button
              class="btn_action"
              id="button-login"
              type="submit"
            >
              Войти
            </button>
          </form>
        </div>
      </section>
    </div>
  </body>,
  "container": <div>
    <header
      class="header"
    >
      <a
        href="/"
      >
        <img
          alt="logo"
          class="logo_big"
          src="/src/assets/images/logo.png"
        />
      </a>
    </header>
    <section
      class="form login"
    >
      <h2
        class="form__title"
      >
        Добро пожаловать 
        <br />
         На Борт Космической Одиссеи!
      </h2>
      <div
        class="form__content"
      >
        <div
          class="form__question"
        >
          Путешествуете с нами впервые? 
          <a
            href="/registration"
          >
            Зарегистрируйтесь!
          </a>
        </div>
        <form
          class="form__inputs"
        >
          <div
            class="form__inputs-wrapper"
          >
            <label
              class="placeholder"
              for="name"
            >
              <input
                class="form__input"
                id="email"
                required=""
                type="text"
              />
              <div
                class="placeholder__input form_big-first-letter"
              >
                Email
                <span>
                  *
                </span>
              </div>
              <p
                class="error-message "
              />
            </label>
          </div>
          <div
            class="form__inputs-wrapper"
          >
            <label
              class="placeholder"
              for="password"
            >
              <input
                class="form__input"
                id="password"
                required=""
                type="password"
              />
              <div
                class="placeholder__input form_big-first-letter"
              >
                Password
                <span>
                  *
                </span>
              </div>
              <button
                class="password-view password-view_password"
                type="button"
              >
                 
              </button>
              <p
                class="error-message"
              />
            </label>
          </div>
          <button
            class="btn_action"
            id="button-login"
            type="submit"
          >
            Войти
          </button>
        </form>
      </div>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
